diff --git a/field_collection.install b/field_collection.install
index 2c3c427..6d53311 100644
--- a/field_collection.install
+++ b/field_collection.install
@@ -280,3 +280,23 @@ function field_collection_update_7004() {
     }
   }
 }
+
+/**
+ * Force the creation of the table cache_entity_field_collection_item.
+ *
+ * entity_update_7003 will attempt to install entitycache tables for existing
+ * modules, but it uses module_list() to get the list of available modules,
+ * which, when called from a database update, may not return field_collection
+ * since drupal is bootstrapped at a lower level.
+ */
+function field_collection_update_7005() {
+  if (module_exists('entitycache')) {
+    $entity_type = 'field_collection_item';
+    $table = 'cache_entity_' . $entity_type;
+    if (!db_table_exists($table)) {
+      $schema = drupal_get_schema_unprocessed('system', 'cache');
+      $schema['description'] = 'Cache table used to store' . $entity_type . ' entity records.';
+      db_create_table($table, $schema);
+    }
+  }
+}
